plutus-core/plutus-core/src/PlutusCore/Compiler/Types.hs

```diff
@@ -6,5 +6,14 @@ import PlutusCore.Builtin
 import PlutusCore.Name
 import PlutusCore.Quote
 
-type Compiling m uni fun name =
-  (ToBuiltinMeaning uni fun, MonadQuote m, HasUnique name TermUnique, Ord name)
+import Data.Hashable
+
+type Compiling m uni fun name a =
+  ( ToBuiltinMeaning uni fun
+  , MonadQuote m
+  , HasUnique name TermUnique
+  , Ord name
+  , Typeable name
```

Comment:
As I commented on your other PR, I don't see why we can't just have a `Hashable name` constraint?

Reply:
Probably you are right; let me try and see if it works.

Reply:
Oh, I remember why I didn't do that: the current implementation of CSE only works correctly for `Name`. We can't just do the substitutions that we do for `DeBruijn` terms.

Reply:
HasUnique?